Include more digestible bits for the introduction of the book

Wilker-uwu opened the pull request:
As discussed in the Discord server, the main problem with the tutorials and introductions included within the book is that it is hard to even grasp what even is an Extension in the context of Godot. Going back to read it a couple of times, the culprit for that is that it includes too much of the preparation and almost none of the process. Basically a lot of "You have to do that", "Do this before you begin using", "This is the difference compared to a version you might have never used", but too little of "This is how it works, here's how to set it up". 

While sure, assuming the reader might have the basic or or moderate understanding of the Rust language and its tools is reasonable to get the message delivered quickly, that does not exempt the book from the necessity of splitting the seemingly obvious details of the process to enable the explanation of its nuances, so keeping variations of "x is y+1 is 11" and having its variations as but a second thought will more often than not wrap it around into complicating it all again, both the learning process as well as the troubleshooting process.

This series of commits is aimed at easing the learning curve to new readers, both those new to Godot in general as well as those who may have just started learning Rust. The ideal scenario is that it will succeed in making a portion of the necessary information understandable at a glance for those who would prefer to scroll faster than reasonable. If all goes well, this should help other people to improve the explanations of the nuances that may affect the inclusion of the library into new and ongoing projects.

Thank you for reading :heart:

###### ps: yes it does have a diagram in it. sorry @Bromeon

Wilker-uwu committed:
Include more digestible bits for the introduction of the book

Bromeon commented on book.toml:
```diff
@@ -15,11 +15,15 @@ renderer = ["html"]
 
 [preprocessor.admonish]
 command = "mdbook-admonish"
-assets_version = "2.0.2" # do not edit: managed by `mdbook-admonish install`
```
Could you keep unrelated config changes out of this PR? This makes CI fail at the moment.

Wilker-uwu replied:
no problem. i just assumed it was okay to commit together because of all the version errors i got when i first started building from source.

Bromeon commented on book.toml:
```diff
@@ -15,11 +15,15 @@ renderer = ["html"]
 [output.html]
 git-repository-url = "https://github.com/godot-rust/book"
-additional-css = ["config/mdbook-admonish.css"]
+additional-css = ["config/mdbook-admonish.css", "./mdbook-admonish.css"]
```
Also here, this shouldn't change.

Bromeon commented on src/index.md:
```diff
@@ -14,15 +14,40 @@ in the officially maintained [Rust Book](https://doc.rust-lang.org/book/).
 
+## The basics
+
+A **library** is in its most basic form is a **collection of functions**. Godot has several of its own built-in libraries, but like most other programs made as a creation tool, it also has the capacity to access collections of libraries made outside of Godot. Those are called **extensions**, and like most other software, they usually come in separate files -- typically in the form of file extensi ...
```
From Discord:
> I would probably try to be strict with terminology, otherwise we might confuse readers rather than clear up:
> - "Library" should be used in the C sense, as static or dynamic library. I don't think for the chapter it's relevant that Godot itself uses libraries (virtually any software does). So I'd focus on the dynamic library -- the GDExtension binding.
> - "Collections" almost always refers to containers and data structures. 

Also twice "is" in first sentence.

Last sentence, "[extensions] come in separate files -- typically in the form of file extensions" sounds weird. The "like most other software, they usually come in separate files" contains not much relevant information, I'd just say something like:

> Those are called **extensions** and are distributed as dynamic libraries, with file endings `.so` on Linux, `.dll` on Windows or `.dylib` on macOS.

Wilker-uwu replied:
i didn't consider that as much of a problem because it's way too common where a software will have a separate file as an integral part of it.

i'm not as sure about the context of Godot in particular which is one binary, but "extensions" would otherwise imply that most other programs don't necessarily need them to have core parts functioning (it would make sense to call FFmpeg an extension for ShareX since it doesn't require for a core function, but doesn't make sense for KDEnLive which requires it for rendering.) thus, being dynamic or not won't make much difference unless the mention here is better to understand the idea. `hello-world.md` and the glossary in this file has a lot of explanation already, so i'm not sure what to do.

i changed `collections` into `packages` though. `package` has a mention in `hello-world`, so i think it makes sense.

Bromeon commented on src/index.md:
```diff
@@ -14,15 +14,40 @@ in the officially maintained [Rust Book](https://doc.rust-lang.org/book/).
+
+A **library** is in its most basic form is a **collection of functions**. Godot has several of its own built-in libraries, but like most other programs made as a creation tool, it also has the capacity to access collections of libraries made outside of Godot. Those are called **extensions**, and like most other software, they usually come in separate files -- typically in the form of file extensi ...
+
+The way that GDExtension work is that you basically expose these files to your Godot project by using a `.gdextension` file pointing to the thing. Such file will have all the information of where that file is and which version of Godot it should be compatible with.
```
> The way that GDExtension work is that you basically expose [...] pointing to the thing.

This sounds quite sloppy; I would avoid unnecessarily vague terms like "thing" (or maybe even "basically") and directly describe what happens, in a concise way:

> You can expose your extension to Godot using a `.gdextension` file, which contains information about the dynamic library's location and the minimum required Godot version.

Wilker-uwu replied:
makes sense.

Bromeon commented on src/index.md:
```diff
@@ -14,15 +14,40 @@ in the officially maintained [Rust Book](https://doc.rust-lang.org/book/).
 ## Terminology
 
-To avoid confusion, here is an explanation of names and technologies you may encounter over time.
+To avoid confusion, here is an explanation of names and technologies you may encounter over time. This is what we will be referring to at the mention of the terms listed from here now on:
```
I don't see what the 2nd sentence adds. I would avoid unnecessary repetition to decrease reading time.

Wilker-uwu replied:
should be shortened now. probably better.

Bromeon commented on src/index.md:
```diff
@@ -14,15 +14,40 @@ in the officially maintained [Rust Book](https://doc.rust-lang.org/book/).
-- **gdnative** (lowercase): the Rust binding for GDNative (Godot 3).
+- [**gdext**] (lowercase): the Rust binding for GDExtension (Godot 4) -- what this book focuses on.
+- [**gdnative**] (lowercase): the Rust binding for GDNative (Godot 3).
 - **Extension**: An extension is a C library developed using gdext. It can be loaded by Godot 4.
```
Nice additions, thanks! 🙂

Wilker-uwu replied:
np :3

(sneakily hides the rickroll link away from "Extension")

Bromeon commented on src/index.md:
```diff
@@ -14,15 +14,40 @@ in the officially maintained [Rust Book](https://doc.rust-lang.org/book/).
+
+The way that GDExtension work is that you basically expose these files to your Godot project by using a `.gdextension` file pointing to the thing. Such file will have all the information of where that file is and which version of Godot it should be compatible with.
+
+The job of a binding library is that we go the other way and we instead treat Godot itself like a library, so that we can have most of the interactions of the engine happen in our language of choice. In practice,  Godot is still going to be the host engine of our game, so we have to expose things back to Godot so that it can use and make available in the editor -- think of how you use the [`@expo ...
```
The "using like a library" is OK, but I don't think going into `@export` is helpful here -- first, it doesn't really elaborate the library concept, and second, you have to assume that the reader is not familiar with exports at this point of reading. The last sentence is also not adding much.

I would just mention two things in this paragraph:

* Godot is the host (runs the process), gdext + user's Rust code is the library
* By using gdext, we can treat Godot like a library and interact with its API

Wilker-uwu replied:
this part was a little awkward to write. got it a little shorter and removed the `@export` bit now.

Bromeon commented on src/intro/objects.md:
````diff
@@ -61,18 +61,21 @@ In particular, its responsibilities include:
 3. A signal handler for the `body_entered` signal of a `Node3D` in your custom class
+
     ```rust
     #[godot_api]
````
Were these empty lines required (CI passed without them before)? Make sure you use the same markdownlint config to avoid unnecessary formatting changes.

Minor thing, but it might turn the list from a "sparse list" into a "spread list", adding `<p>` tags in HTML. Maybe it has no effect though.

Wilker-uwu replied:
i don't remember touching this file honestly. might be a leftover of the linting i forgot. one of its settings is in regards to "all code blocks should be spaced". i'm reverting this change anyway.

Bromeon commented on src/intro/setup.md:
````diff
@@ -17,16 +17,55 @@ For the rest of the tutorial, we assume that you have Godot 4 installed and avai
+brew install godot
+```
+
+NOTE: If you plan to target Godot versions different from the latest stable release, please read [Compatibility and stability][compatibility].
````
Please use admonish block.

Wilker-uwu replied:
included now.

Bromeon commented on src/intro/setup.md:
```diff
@@ -17,16 +17,55 @@ For the rest of the tutorial, we assume that you have Godot 4 installed and avai
 Binaries of Godot 4 can be downloaded [from the official website][godot-download].
 
-If you plan to target Godot versions different from the latest stable release, please read [Compatibility and stability][compatibility].
+### Installing Godot through commands
```
Maybe `### Pre-built binaries` and `### Installing via command-line` or so?

Wilker-uwu replied:
there's more than one program listed in this page, so i'm keeping the name with each header. see if this is okay

Bromeon commented on src/intro/setup.md:
```diff
@@ -17,16 +17,55 @@ For the rest of the tutorial, we assume that you have Godot 4 installed and avai
+winget install --id=GodotEngine.GodotEngine -e
+
+
+#-- MacOS --#
```
Note on capitalization: should be `macOS`.

Also in other places.

Wilker-uwu replied:
now changed 2 other entries i think, found and ignored 1 in a file i don't know.

Bromeon commented on src/intro/setup.md:
```diff
@@ -64,9 +108,10 @@ clang -v
 
 [compatibility]: ../advanced/compatibility.md
-[godot-download]: https://godotengine.org/download
+[godot-download]: https://godotengine.org/download/archive/
```
Maybe the archive should be listed separately from the download page, in parentheses.

Wilker-uwu replied:
tried a short phrasing.

Bromeon commented on src/intro/setup.md:
```diff
@@ -64,9 +108,10 @@ clang -v
 [llvm]: https://releases.llvm.org
 [rustup-windows]: https://github.com/rust-lang/rustup#working-with-rust-on-windows
 [rustup]: https://rustup.rs
+[`winget`]: https://github.com/microsoft/winget-cli
```
This seems unreferenced.

Wilker-uwu replied:
for me it's 111 and 119 that's unreferenced for me. this linting is weird x.x

Bromeon commented on src/resources/images/godot_project_creation.png:
Thanks for this image! 🙂 

Is it possible to decrease the file size a bit, e.g. by having a blank background?

Keep in mind that Git repositories are ever-growing, so anything ever committed will need to be downloaded by every single user until the end of time.

Wilker-uwu replied:
i squished the picture, it is now 81.5 KB, down from 331.7 KB. not sure how Git works to know if it will matter though.

Bromeon commented on src/resources/images/scripting_first_script_godot_turning_in_place.gif:
I think the GIFs can keep being referenced externally.

Bromeon reviewed (commented):
Thanks a lot! I added quite a few comments, but most of them focus on formal things rather than content, so overall it's a great contribution 😊

Maybe some general advice: while it may sound boring at times, the book is primarily a technical documentation, so descriptions should be unambiguous and concise. I would avoid phrases that don't add information, as they increase reading time and make it harder to find information later on.

I'm still not convinced we need the diagram here 😉 what about the concise architecture one I suggested on Discord?

Wilker-uwu commented on src/intro/setup.md:
```diff
@@ -17,16 +17,55 @@ For the rest of the tutorial, we assume that you have Godot 4 installed and avai
 
-If you plan to target Godot versions different from the latest stable release, please read [Compatibility and stability][compatibility].
+### Installing Godot through commands
+
```
i took rent.

Wilker-uwu commented on src/intro/hello-world.md:
```diff
@@ -69,39 +164,19 @@ res://HelloWorld.gdextension
 
-Our C library needs to expose an _entry point_ to Godot: a C function that can be called through the GDExtension.
-Setting this up requires quite some low-level FFI code, which is why gdext abstracts it for you.
+[Like mentioned earlier](#why-cdylib), our compiled C library needs to expose an _entry point_ to Godot: a C function that can be called through the GDExtension.
```
the linting here screams at me for using `a code mark` in a header. no idea how to fix it.

Wilker-uwu commented on src/intro/hello-world.md:
```diff
@@ -326,9 +444,16 @@ API attributes typically follow the GDScript keyword names: `class`, `func`, `si
 
 
-[derive-godotclass]: https://godot-rust.github.io/docs/gdext/master/godot/prelude/derive.GodotClass.html
 [godot-command-line]: https://docs.godotengine.org/en/stable/tutorials/editor/command_line_tutorial.html
```
linting marks first and last links to be unused when they're clearly being used, not sure how to fix this one either.

Bromeon commented on src/index.md:
```diff
@@ -8,21 +8,43 @@
+## The basics
+
+A **library** in its most basic form is a **package of functions**. Godot has several of its own built-in libraries, but like most other programs made
+as a creation tool, it also has the capacity to access collections of libraries made outside of Godot. Those are called **extensions**, and like most
```
Again: I would really stay on-topic. It's not relevant for the understanding that:
* Godot uses its own built-in libraries (because every non-hello-world software on this planet does this)
* Creation tools in general use other collections of libraries.

What is relevant is that Godot provides a mechanism to extend its own functionality through dynamic libraries.

Wilker-uwu replied:
the main idea here was an explanation by comparison. although technical, it's probably a good idea to still have a few outside context here and there. i'd keep it since it's not as distracting as some of the other parts you pointed out.

------
maybe a minor wording change, but otherwise i'm out of ideas for this bit:

```suggestion
A **library** in its most basic form is a **package of functions**. Godot has several of its own built-in libraries, but like most other programs made
as a creation tool, it also has the capacity to extend its functionality by accessing collections of libraries made outside of Godot. Those are
```

Bromeon commented on src/index.md:
```diff
@@ -8,21 +8,43 @@
+You can expose your extension to Godot using a `.gdextension` file, which contains information about the dynamic library file's location and the
+minimum required Godot version.
+
+The job of a binding library is that we go the other way and we instead treat Godot itself like a library, so that we can have most of the
```
Maybe use "language binding" instead of "binding library" here.

Wilker-uwu replied:
change made.

Bromeon commented on src/index.md:
```diff
@@ -8,21 +8,43 @@
+to Godot so that it can use and make available in the editor.
+
+gdext bridges Godot functions we can call as a library, then we build and pass it back to Godot. Godot uses all that we wrote and uses to run what we
+need into the game.
```
I'm not really sure I follow here. But I also don't think it's necessary to add more information on this, a library already implies that you can call functions and expose code. So I'd remove it.

Wilker-uwu replied:
that's fair. i removed the bit then.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+- A Cargo crate of your own;
+- A Godot project created;
+
+You can very much start from an already existing Godot project. Here we'll start from scratch for the sake of completion.
```
"completeness", not "completion" 🙂

Wilker-uwu replied:
fixed now :3

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+For the following objective, you are going to need
+
+- A Cargo crate of your own;
+- A Godot project created;
```
This reads like the user already needs to set these things up -- however you explain in the next 2 sections to do exactly that. 

Rather formulate it in a way that the following sections are going to explain that, not that the user needs it.

Wilker-uwu replied:
```suggestion
First things first.  
For the following objective, we'll first make

- A Cargo crate of your own;
- A Godot project created;
- ~~me cry~~
```

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+### Creating the Godot Project
+
+For this example we're going to create each one separated into neighboring folders in the same parent folder. You'll be able to choose locations of
+your own as we progress. Let's start by creating a blank Godot project if you haven't already.
```
What does "each one" refer to?

Make sure it's completely unambiguous what you refer to.

Wilker-uwu replied:
pretty sure "each of the items" should be clear enough.
```suggestion
For this example we're going to create each of the items separated into neighboring folders in the same parent folder. [...]
```

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+For this example we're going to create each one separated into neighboring folders in the same parent folder. You'll be able to choose locations of
+your own as we progress. Let's start by creating a blank Godot project if you haven't already.
+
+![alt: A screenshot of the GUI of the Godot project manager about to create a new project][img-godot-create-project]
```
No need for `alt: ` prefix.

Wilker-uwu replied:
not sure how to format this, this often becomes indistinguishable from normal text whenever the image fails to load, becomes distracting to read.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+
+### Creating the Rust Cargo crate
+
+To create a Cargo crate, open have your terminal open, navigate to your desired folder and then call `cargo` to create it by typing
```
Twice "open".

Also end with a colon.

Wilker-uwu replied:
fixed.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+
+The argument `--lib` will tell `cargo` to create the package from a Library template. From there, we're going to configure it in the file which stores
+its metadata, `Cargo.toml`. This file stores a lot of the information we know of what the library package is about and what it uses in order to make
+it work.
```
> This file stores a lot of the information we know of what the library package is about and what it uses in order to make it work.

This sentence is overly complex and needs to be re-read multiple times to be understood. Phrases like "we know of", "what it is about", "what it uses to make it work" etc. do not really add anything but rather obfuscate the meaning.

I would omit the whole sentence -- in general, this is not the place to explain what `Cargo.toml` is. Users are expected to have a basic understanding of Rust or at least read the Rust book in parallel.

Wilker-uwu replied:
better keep it for completeness. this is what this whole section of the file is about.
```suggestion
This file stores the information we know of the library package and what it uses in order to work with it.
```

Bromeon commented on src/intro/hello-world.md:
````diff
@@ -4,104 +4,203 @@
+
+```toml
+[package]
+name = "rust_project" # this name will affect the name of the library file
````
What is the "library file"? Be precise, otherwise it's mistaken with the `.gdextension` file.

```suggestion
name = "rust_project" # Will appear in the filename of the dynamic Rust library you compile.
```

Wilker-uwu replied:
makes sense. will pick it up.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+edition = "2021" # Rust edition version (remains unchanged)
+
+[lib]
+crate-type = ["cdylib"] # defines the kind of library it will become
```
```suggestion
crate-type = ["cdylib"] # Compile this crate to a dynamic C library.
```

Wilker-uwu replied:
this was intended to describe the keys, not the values.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+[lib]
+crate-type = ["cdylib"] # defines the kind of library it will become
+
+[dependencies] # lists what your own library needs in order to function
```
This is obvious, no comment needed here.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
-Once you have a Godot project, you can register GDExtension libraries that it should use. These libraries are called _extensions_
-and loaded during Godot engine startup -- both when opening the editor and when launching your game. The registration requires two files.
+The `cdylib` crate type is a bit unusual in Rust. Instead of building a binary (`bin`, application) or a library to be utilized by other Rust code
+(`lib`), we create a _dynamic_ library, exposing an interface in the C programming language This dynamic library is loaded by Godot at runtime,
```
Missing full stop after "programming language".

Wilker-uwu replied:
fixed.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
-Save the changes and consider versioning each step of the tutorial in Git.
+#### Why "`cdylib`"?
+<!-- editors note: sole reason this header's here is because this marks a quick spot to go back to later on,
+specially if a new reader is just scrolling by around the book. -->
```
Makes sense; this could actually be an admonish note.

I would also recommend to give titles to individual notes:
```md
```admonish note title="cdylib"
Text here...
```

Also applies to other notes.

Wilker-uwu replied:
couldn't fit a pair of quotation marks and another of crases (this thing here ` idk the english name of) within the title. it was another point where i just gave up rather than compromising on writing somethine else entirely. ("why cdylib?" vs. just "cdylib")

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+│ 3 │ target/debug/incremental        │ dir  │    52 B │ 16 seconds ago │
+│ 4 │ target/debug/librust_project.d  │ file │   190 B │ now            │
+│ ⮞ │ target/debug/librust_project.so │ file │ 62.6 MB │ now            │
+╰───┴─────────────────────────────────┴──────┴─────────┴────────────────╯
```
Where is this from? As far as I know, it doesn't appear automatically upon `cargo build`.

Wilker-uwu replied:
`ls` command output from the Nu Shell, included for visual clarity. added the extra line to avoid confusion.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
 
-First, add `res://HelloWorld.gdextension`, which is the equivalent of `.gdnlib` for GDNative.
+The job of this file is to provide Godot with pointers to a library that is compiled elsewhere. We're going to use it to connect Godot to your
+compiled library so it can be used within the context of your Godot project.
```
```suggestion
This file tells Godot how to load your compiled Rust extension: the path to the dynamic library file, and the 
entry point (function) to initialize it with.
```

Wilker-uwu replied:
the phrasing feels easier to grasp as it is now. explanation bringing up the terms already shows up later on.

Bromeon commented on src/intro/hello-world.md:
````diff
@@ -4,104 +4,203 @@
+```
+
+The `..` is a pointer to a parent folder of the end path.
+Which means Godot will look for the library around the parent folder of the project for its path.
````
Can also be simplified:
```suggestion
The `..` refers to a parent folder, i.e. Godot will look in a path relative to the project directory's parent.
```

Wilker-uwu replied:
a more direct representation:
```suggestion
The `..` is a pointer to a parent folder of the end path.
e.g Godot will look in a path relative to this project directory's parent folder.
```

Bromeon commented on src/intro/hello-world.md:
````diff
@@ -4,104 +4,203 @@
+Which means Godot will look for the library around the parent folder of the project for its path.
+
+```admonish tip
+You can also employ the use of synbolic links and git modules and then treat those as regular folders and files. Godot reads those just fine too! 
````
Nice addition!

You probably meant "symbolic" and "git submodules"?

Wilker-uwu replied:
fixed :3

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -4,104 +4,203 @@
+---
 
-In your terminal, create a new Cargo project with your chosen crate name:
+From here on out, your Rust project files should be connected to Godot, so now we'll make it actually function.
```
The line is not needed.

"actually function" is again imprecise. What is left?
It's also fine to move this to the next section without announcing it beforehand.

Wilker-uwu replied:
i admit i just wanted to humor a bit with this line. see also: https://www.youtube.com/watch?v=-A5B6jqtt-c

Bromeon commented on src/intro/hello-world.md:
````diff
@@ -4,104 +4,203 @@
-```
 
-Then, open Cargo.toml and add the following:
+## Making Rust actually function
````
Describe _what_ is being set up here.

Wilker-uwu replied:
no fun allowed :c

but fair.

but also this feels like it could use an open-ended title like it.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -326,9 +472,18 @@ API attributes typically follow the GDScript keyword names: `class`, `func`, `si
+"A screenshot of the GUI of the Godot project manager about to create a new project"
+
+
+[^ffi]: FFI: [Foreign Function Interface](https://en.wikipedia.org/wiki/Foreign_function_interface)
```
This could be a regular link, no need for a footnote.

Wilker-uwu replied:
a quick scroll down for a term that's never used again in this page, rather than an external link. but that's fair. change made.

found out alt text also works in links. cool.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -114,24 +213,70 @@ unsafe impl ExtensionLibrary for MyExtension {}
 
-1. Import the `prelude` module from the `godot` crate. This module contains the most common symbols in the gdext API.
+1. Place the [`prelude`][api-prelude] module from the [`godot`][api-godot] crate into scope.
+   This module contains the most common symbols in the gdext API.
 2. Define a struct called `MyExtension`. This is just a type tag without data or methods, you can name it however you like.
-3. Implement the `ExtensionLibrary` trait for our type, and mark it with the `#[gdextension]` attribute.
+3. Implement the [`ExtensionLibrary`][api-extensionlibrary] trait for our type, and mark it with the `#[gdextension]` attribute.
```
Thanks a lot for these links! 👍

Wilker-uwu replied:
np :3

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -114,24 +213,70 @@ unsafe impl ExtensionLibrary for MyExtension {}
+#### "Does it load though?"
+
+If you want a quick look at having it load, you can start out by overloading the `on_level_init(InitLevel)` from the
+[`ExtensionLibrary`][api-extensionlibrary] trait like so:
```
```suggestion
If you want to see Godot loading your extension, you can override `on_level_init(InitLevel)` from the
[`ExtensionLibrary`][api-extensionlibrary] trait like so:
```

Wilker-uwu replied:
done.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -114,24 +213,70 @@ unsafe impl ExtensionLibrary for MyExtension {}
 
 
-## Creating a Rust class
+#### "Does it load though?"
```
```suggestion
### Loading the extension
```

(Level 3 should be OK)

Wilker-uwu replied:
this phrasing implies that loading the extension requires that `on_level_init` is overloaded. i wanted to clarify the intent by having it as a minor section instead

Bromeon commented on src/intro/hello-world.md:
````diff
@@ -114,24 +213,70 @@ unsafe impl ExtensionLibrary for MyExtension {}
+}
+```
+
+And then after `cargo build`, Godot should display the messages when you run Godot from the command line
````
```suggestion
After `cargo build`, Godot should display the following messages when running it from the command line:
```

Wilker-uwu replied:
meant to write "these", less verbose. still picking your suggestion though.

```suggestion
After `cargo build`, Godot should display these messages when you run Godot from the command line
```

Bromeon commented on src/intro/hello-world.md:
````diff
@@ -114,24 +213,70 @@ unsafe impl ExtensionLibrary for MyExtension {}
+```admonish note
+These messages will most likely not show up unless you're either launching [Godot editor from your command line][godot-command-line] by using
+`godot -e` like in this example, or making your gdextension file visible for the first time in Godot while opened.
+```
````
Instead of this note, we should probably explain the role of the `.godot/extension_list.cfg` file. Once that is there, this is not a problem and there's no probability involved anymore.

Wilker-uwu replied:
i'm learning to use this whole project as i'm writing this, and the extension_list file is actually one of the things i don't know how to explain yet ^^' this bit was an observation from asking questions in the Discord (and trying stuff on my own).

there is a point up in this file briefly mentioning it exists. maybe an explanation there would be fitting?

i was actually considering placing this whole section inside a collapsed version of the Admonish Example block since this is a section that i could see people ignoring without missing much. by "considering" i mean i failed to group it for an hour before i gave up.

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -150,42 +295,44 @@ struct Player {
 Let's break this down.
 
-1. The gdext prelude contains the most common symbols. Less frequent classes are located in the `engine` module.
+1. We define two fields `speed` and `angular_speed` for the logic. These are regular Rust fields, no magic involved. More about their use later.
```
The previous order was better, since it followed the order of different symbols appearing in the code.

I appreciate your improvements though (links + formulation), so would just need swapping some numbers 🙂

Wilker-uwu replied:
sure ^^

Bromeon commented on src/intro/hello-world.md:
```diff
@@ -150,42 +295,44 @@ struct Player {
    native inheritance).
 
    - The field must have type `Base<T>`.
+     - The type `T` considered valid is affected by what class you've chosen in `#[class(base=Sprite2D)]`
```
Good addition! I would maybe write it a bit more concretely:
```suggestion
     - `T` must match the declared base class, e.g. `Sprite2D` for `#[class(base=Sprite2D)]`.
```

Wilker-uwu replied:
also included "becomes `Base<Sprite2D>`" just in case.
```suggestion
     - `T` must match the declared base class, e.g. `Sprite2D` for `#[class(base=Sprite2D)]` becomes `Base<Sprite2D>`.
```

Bromeon commented on src/intro/setup.md:
```diff
@@ -17,16 +17,60 @@ For the rest of the tutorial, we assume that you have Godot 4 installed and avai
 ## Rust
 
-[rustup] is the recommended way to install the Rust toolchain, including the compiler, standard library and Cargo (the package manager).
+[rustup][rustup] is the recommended way to install the Rust toolchain, including the compiler, standard library and Cargo (the package manager).
```
No need to repeat the link if the referred-to identifier already matches the one in `[ ]` (that's why it wasn't done before).

Wilker-uwu replied:
sure

Bromeon commented on src/resources/images/godot_project_creation-compressed-more.png:
As mentioned previously, you could just use a blank Desktop background. Not only would this distract less from the important parts (the window), but it would need much less size while still having decent resolution for the parts that matter. Now, the title bars are quite blurry for example.

If it's too much work, don't worry, it's fine to omit this. There's really not much a user can do wrong here, the menu is super intuitive.

Wilker-uwu replied:
it's not blurry whenever i open this outside of a browser.

i hate bilinear filter so much.

Bromeon reviewed (commented):
Thanks for the updates! It's slowly taking shape 😀 

Made more comments. Sorry if I'm sometimes too nitpicky on the writing style, but it's really crucial to me that everything is written as clearly and unambiguously as possible, otherwise we'll just end up with frustrated users and more support questions on Discord and GitHub.

---

Also, I saw you changed a lot of line widths in sections that are unrelated to your additions. Have they been introduced by automated tooling? It's not bad to have things consistent, but now it's quite a bit harder to see what is new.

As a rule of thumb, if it passes markdownlint CI, the formatting should be fine. For now you can leave it though, no need to revert 🙂 

---

The images and diagrams are also still there. If you have the opportunity, could you respond to [my request](https://github.com/godot-rust/book/pull/9#pullrequestreview-1688651661) about the architecture diagram and remove everything that's not needed?

Wilker-uwu reviewed (commented):
pwease wewiew

(i think this finishes my review then? took a bit to find the button on the github page)

Bromeon committed:
Follow-up improvements

Mostly points from my own review + a bit of additional content to make starting easier.
